File: src/ee/core/internal/Thread.swift
//
//  Thread.swift
//  Pods
//
//  Created by eps on 6/17/20.
//

import Foundation

@objc(EEThread)
public class Thread: NSObject {
    @objc
    public class func isMainThread() -> Bool {
        return Foundation.Thread.isMainThread
    }

    @objc
    public class func runOnMainThread(_ callback: @escaping () -> Void) -> Bool {
        if isMainThread() {
            callback()
            return true
        }
        DispatchQueue.main.async {
            callback()
        }
        return false
    }

    @objc
    public class func runOnMainThreadDelayed(_ seconds: Float,
                                             _ callback: @escaping () -> Void) {
        DispatchQueue.main.asyncAfter(
            deadline: .now() + .milliseconds(Int(seconds * 1000))) {
            callback()
        }
    }
}

@_cdecl("ee_isMainThread")
public func ee_isMainThread() -> Bool {
    return Thread.isMainThread()
}

@_cdecl("ee_runOnMainThread")
public func ee_runOnMainThread() -> Bool {
    return Thread.runOnMainThread {
        ee_runOnMainThreadCallback()
    }
}

@_cdecl("ee_runOnMainThreadDelayed")
public func ee_runOnMainThreadDelayed(_ key: Int32, delay: Float) {
    Thread.runOnMainThreadDelayed(delay) {
        ee_runOnMainThreadDelayedCallback(key)
    }
}

File: proj.android_studio/ee-x-core/src/main/java/com/ee/Thread.kt
package com.ee

import androidx.annotation.AnyThread
import com.ee.internal.IThreadImpl

object Thread {
    private var _impl: IThreadImpl? = null

    @AnyThread
    fun registerImpl(impl: IThreadImpl) {
        _impl = impl
    }

    @AnyThread
    @JvmStatic
    fun checkMainThread() {
        _impl?.checkMainThread()
    }

    @AnyThread
    @JvmStatic // Used by UnityThread.
    fun isMainThread(): Boolean {
        return _impl?.isMainThread() ?: false
    }

    @AnyThread
    @JvmStatic // Used by UnityThread.
    fun runOnMainThread(callback: Runnable): Boolean {
        return _impl?.runOnMainThread(callback) ?: false
    }

    @AnyThread
    @JvmStatic
    fun runOnMainThreadDelayed(seconds: Float, callback: Runnable) {
        _impl?.runOnMainThreadDelayed(seconds, callback)
    }
}

private external fun ee_runOnMainThreadCallback()
private external fun ee_runOnMainThreadDelayedCallback(key: Int)

@Suppress("unused")
private fun ee_isMainThread(): Boolean {
    return Thread.isMainThread()
}

@Suppress("unused")
private fun ee_runOnMainThread(): Boolean {
    return Thread.runOnMainThread {
        ee_runOnMainThreadCallback()
    }
}

@Suppress("unused")
private fun ee_runOnMainThreadDelayed(key: Int, delay: Float) {
    Thread.runOnMainThreadDelayed(delay) {
        ee_runOnMainThreadDelayedCallback(key)
    }
}